### ISSUE TITLE
Replace ratings in DB by ratings stored in files

### ISSUE THREAD

**Issue by Clementine-Issue-Importer** [clementine-player/Clementine#3666]
_From [alphadel...@gmail.com](https://code.google.com/u/107136459110240960141/) on May 10, 2013 19:46:13_

What steps will reproduce the problem? 1.  Change the FMPS_Rating tag in a file.
2.  Check the file's rating in Clementine.
3.  Updated changed library folders in Clementine.

What is the expected output?
The song's rating should be updated in Clementine.

What do you see instead?
It is not. What version of the product are you using? On what operating system? 1.1.1-498-ge456ad6~quantal Please provide any additional information below. I mounted my laptop's ~/Music folder on my desktop over sshfs and used Clementine on my desktop to change a file's rating.  The FMPS_Rating tag was changed, and the file's mtime was changed, but when I look at the file in Clementine on my laptop, the rating is still 0.  

I even "update[d] changed library folders" but that didn't detect the change.  I notice that the directory's mtime didn't change with the file's being changed, so maybe that's why Clementine didn't notice.

_Original issue: http://code.google.com/p/clementine-player/issues/detail?id=3666_

**Comment by alphapapa** [comment]
Ok, I just pushed https://github.com/alphapapa/Clementine/commit/635b137f1767e40c91b07fce106dded5bdb5e598 which causes changed `FMPS_Rating` tags to be re-read from files.  It was a two-line change.  :)

Now I have noticed a minor bug that may be an old bug, or newly introduced by my change.

1.  Add a song to the library.
2.  Add it to a playlist.
3.  Change the song's rating on-disk (not in Clementine).
4.  Open the song's "Edit track info" dialog.
5.  See that the new rating is listed, but in the playlist, the old rating is still listed.
6.  Drag the song from the library pane to the playlist again.
7.  See the old rating still displayed in the playlist.
8.  Restart Clementine.
9.  See the old rating still displayed in the playlist.
10.  Open the song's "Edit track info" dialog again and see that the new rating is still displayed.
11. Tools->Update changed library folders
12.  Drag song from library to playlist again.
13.  Old rating still displayed in playlist.
14.  Tools->Do a full library rescan
15.  The new rating is immediately displayed in the playlist.
16.  Drag song from library to playlist again, and new rating is also correctly displayed.

I am guessing that this is a pre-existing bug, not introduced by my change, but I'm not sure.  What do you think?

If it's a separate bug, will you accept a pull request for this change?

Thanks.

**Comment by ArnaudBienner** [comment]
@alphapapa if I understand well, then it is probably introduced by your changes indeed: previously, we never read the rating again once the song was indexed.
So, that would mean your patch isn't completely working? If modifying the song doesn't update it within Clementine until you do a full rescan, probably something is missing.
Do you have the problem with other tags? (I can't test right now)

**Comment by alphapapa** [comment]
Thanks for your reply.  :)

I am far from being familiar enough with the Clementine codebase to say for certain, but I don't think that this bug is caused by my two-line patch.  As I noted in point 5 above, the changed rating is picked up by Clementine, because it's visible in the track's "Edit track info" dialog.  Also, in the debug output, I see Clementine picking up the change.

The problem is that the changed rating is not reflected in playlists.

I can think of a couple of possibilities:

1.  The "Edit track info" dialog is reading the rating from the file instead of the database.  That would seem like a bug in and of itself.  If this were the case, it could also mean that the database is not being updated with the changed rating.  But as I said, from looking at the debug output, this doesn't appear to be the case.  Having said that, it's been over 3 weeks since I worked on this, so I'll need to look into it again to refresh my memory.

2.  The playlists are loading track ratings from some kind of cache instead of from the updated database entries, and doing a full refresh update.  Again, I am not familiar enough with the code to know if this is the case, so I'm just speculating.

>  Do you have the problem with other tags?

Good question.  I will have to look into this.  :)

**Comment by svnpenn** [comment]
https://github.com/clementine-player/Clementine/issues/5145

**Comment by svnpenn** [comment]
@hatstand any interest in this? It seems @ArnaudBienner is not interested

**Comment by svnpenn** [comment]
@hatstand any interest in this? It seems @ArnaudBienner is not interested. Also @alphapapa, any updates?

**Comment by svnpenn** [comment]
@alphapapa any updates?

**Comment by alphapapa** [comment]
No, I haven't touched this in quite a while.  As much as I love Clementine, I mostly use MPD now, along with some scripts that adjust tags and ratings from the command line, and Cantata for a GUI.  I still use Clementine sometimes, but not for rating songs, because it's a lot easier to just open the Yakuake window and type "rct 4" for "rate current track 4", and things like that.

I'm not very familiar with programming in Qt, so it would take me a while to figure this out.  For someone that's familiar with Qt and Clementine's code, it would probably be simple to add a few lines of code to update a song's metadata in opened playlists when a song's metadata is updated in the library.

**Comment by svnpenn** [comment]
@hatstand do you know how we might go about fixing this?
